rules/type-declaration/tests/Rector/FunctionLike/ParamTypeDeclarationRector/Fixture/dataprovider_array.php.inc

```diff
@@ -0,0 +1,49 @@
+<?php
+
+namespace Rector\TypeDeclaration\Tests\Rector\FunctionLike\ParamTypeDeclarationRector\Fixture;
+
+/**
+ * Test case including a dataProvider using an array that lead to an incorrect types being added.
+ * `array` was added for $one, but should be `string`, or the parameter should be left alone entirely.
+ * The $two and $three parameters did not get hints added at all.
+ * See: https://phpunit.readthedocs.io/en/9.3/writing-tests-for-phpunit.html#data-providers
+ */
+final class DemoArrayTest
+{
+    public function provideThings(): array {
+        return [
+            [ 123, true, 'I am a string', null ],
+            [ 123, true, 'I am a string', 999 ],
+        ];
+    }
+
+    /**
+     * @dataProvider provideThings
+     */
+    public function testGetFromId( $one, $two, $three, $four ): void {}
+}
+-----
+<?php
+
+namespace Rector\TypeDeclaration\Tests\Rector\FunctionLike\ParamTypeDeclarationRector\Fixture;
+
+/**
+ * Test case including a dataProvider using an array that lead to an incorrect types being added.
+ * `array` was added for $one, but should be `string`, or the parameter should be left alone entirely.
+ * The $two and $three parameters did not get hints added at all.
+ * See: https://phpunit.readthedocs.io/en/9.3/writing-tests-for-phpunit.html#data-providers
+ */
+final class DemoArrayTest
+{
+    public function provideThings(): array {
+        return [
+            [ 123, true, 'I am a string', null ],
+            [ 123, true, 'I am a string', 999 ],
+        ];
+    }
+
+    /**
+     * @dataProvider provideThings
+     */
+    public function testGetFromId( int $one, bool $two, string $three, $four ): void {}
```

Comment:
Very nice work!

Just one detail, looking at the data provider, this should be:

```suggestion
    public function testGetFromId( int $one, bool $two, string $three, ?int $four ): void {}
```

Reply:
I agree, but I also think that could perhaps be fixed in a separate PR.
I could look at chaining that onto this one, similarly to https://github.com/rectorphp/rector/pull/5625 :)
But the fix for data providers here I think is already well worth it as it is!

Reply:
> I agree, but I also think that could perhaps be fixed in a separate PR.
> I could look at chaining that onto this one, similarly to #5625 :)

Good idea :+1: I'll merge this PR as it is, thank you for your wor :)

Reply:
This will end up bring fixed in https://github.com/rectorphp/rector/pull/5659